src/main/java/org/ebyhr/trino/storage/operator/AvroPlugin.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ebyhr.trino.storage.operator;
+
+import io.trino.spi.Page;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.type.Type;
+import org.apache.avro.Schema;
+import org.apache.avro.file.DataFileStream;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+import org.ebyhr.trino.storage.StorageColumnHandle;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+
+import static org.ebyhr.trino.storage.operator.AvroTypeTranslator.fromAvroType;
+
+public class AvroPlugin
+        implements FilePlugin
+{
+    private static final int INITIAL_BATCH_SIZE = 4 * 1024; // 4 KB
+
+    @Override
+    public List<StorageColumnHandle> getFields(String path, Function<String, InputStream> streamProvider)
+    {
+        try (InputStream input = streamProvider.apply(path);
+                DataFileStream<Object> dataFileStream = new DataFileStream<>(input, new GenericDatumReader<>())) {
+            Schema schema = dataFileStream.getSchema();
+            return schema.getFields().stream()
+                    .map(field -> new StorageColumnHandle(
+                            field.name(),
+                            fromAvroType(field.schema())))
+                    .toList();
+        }
+        catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public Iterable<Page> getPagesIterator(String path, List<String> handleColumns, Function<String, InputStream> streamProvider)
+    {
+        try (InputStream input = streamProvider.apply(path);
+                DataFileStream<Object> dataFileStream = new DataFileStream<>(input, new GenericDatumReader<>())) {
+            List<Schema.Field> handledFields = dataFileStream.getSchema().getFields().stream()
+                    .filter(field -> handleColumns.contains(field.name().toLowerCase()))
+                    .toList();
+            // select count(*)
+            if (handleColumns.isEmpty()) {
```

Comment:
The correlation between handleFields and handleColumns is indirect
```suggestion
            if (handleFields.isEmpty()) {
```